RecoParticleFlow/PFClusterProducer/src/PFClusterEMEnergyCorrector.cc

```diff
@@ -116,156 +140,367 @@ void PFClusterEMEnergyCorrector::correctEnergies(const edm::Event &evt, const ed
+      }
+      
+      double invE = (e == 0.) ? 0. : 1./e; //guard against dividing by 0.      
+      int size = lazyTool.n5x5(cluster);      
+      bool iseb = cluster.layer() == PFLayer::ECAL_BARREL;
+      
+      //find index of corrections (0-4 for EB, 5-9 for EE, depending on cluster size and raw pt)
+      int coridx = std::min(size,3)-1;
+      if (coridx==2) {
+	if (pt>4.5) {
+	  coridx += 1;
+	}
+	if (pt>18.) {
+	  coridx += 1;
+	}
+      }
+      if (!iseb) {
+	coridx += 5;
+      }
+      
+      const GBRForestD &meanforest = *forestH_mean[coridx].product();
+      const GBRForestD &sigmaforest = *forestH_sigma[coridx].product();
+      
+      //find seed crystal indices
+      int ietaix=0;
+      int iphiiy=0;
+      if (iseb) {
+	EBDetId ebseed(cluster.seed());
+	ietaix = ebseed.ieta();
+	iphiiy = ebseed.iphi();
+      }
+      else {
+	EEDetId eeseed(cluster.seed());
+	ietaix = eeseed.ix();
+	iphiiy = eeseed.iy();      
+      }
+      
+      
+      //compute preshower energies for endcap clusters
+      double ePS1=0, ePS2=0;
+      if(!iseb) {
+	auto ee_key_val = std::make_pair(idx,edm::Ptr<reco::PFCluster>());
+	const auto clustops = std::equal_range(assoc.begin(),
+					       assoc.end(),
+					       ee_key_val,
+					       sortByKey);
+	for( auto i_ps = clustops.first; i_ps != clustops.second; ++i_ps) {
+	  edm::Ptr<reco::PFCluster> psclus(i_ps->second);
+	  switch( psclus->layer() ) {
+	  case PFLayer::PS1:
+	    ePS1 += psclus->energy();
+	    break;
+	  case PFLayer::PS2:
+	    ePS2 += psclus->energy();
+	    break;
+	  default:
+	    break;
+	  }
+	}
+      }
+      
+      //fill array for forest evaluation
+      eval[0] = evale;
+      eval[1] = ietaix;
+      eval[2] = iphiiy;
+      if (!iseb) {
+	eval[3] = ePS1*invE;
+	eval[4] = ePS2*invE;
+      }
+      
+      //these are the actual BDT responses
+      double rawmean = meanforest.GetResponse(eval.data());
+      double rawsigma = sigmaforest.GetResponse(eval.data());
+      
+      //apply transformation to limited output range (matching the training)
+      double mean = meanoffset + meanscale*vdt::fast_sin(rawmean);
+      double sigma = sigmaoffset + sigmascale*vdt::fast_sin(rawsigma);
+      
+      //regression target is ln(Etrue/Eraw)
+      //so corrected energy is ecor=exp(mean)*e, uncertainty is exp(mean)*eraw*sigma=ecor*sigma
+      double ecor = vdt::fast_exp(mean)*e;
+      double sigmacor = sigma*ecor;
+      
+      cluster.setCorrectedEnergy(ecor);
+      cluster.setCorrectedEnergyUncertainty(sigmacor);
+      
+    }
+    return;
   }
-  else {
-    bunchspacing = bunchSpacingManual_;
+
+  // Selective Readout Flags
+  edm::Handle<EBSrFlagCollection> ebSrFlags;   
+  evt.getByToken(ebSrFlagToken_, ebSrFlags);  
+  if (!ebSrFlags.isValid()){    
+    edm::LogWarning("PFClusterEMEnergyCorrector") << "This version of PFCluster corrections requires the EBSrFlagCollection  information to proceed!!! I will not correct. Will stop here ";
```

Comment:
If this condition happens, it will most certainly repeat in every event.
If it's OK to keep running, we should not flood the logs with LogWarnings and use LogInfo instead.
If it is not OK, this should be an exception.

I think that it should be an exception.